samcli/lib/sync/infra_sync_executor.py

```diff
@@ -0,0 +1,284 @@
+"""
+InfraSyncExecutor class which runs build, package and deploy contexts
+"""
+import logging
+import re
+from typing import Dict, List, Optional, Set
+
+from boto3 import Session
+from botocore.exceptions import ClientError
+
+from samcli.commands._utils.template import get_template_data
+from samcli.commands.build.build_context import BuildContext
+from samcli.commands.deploy.deploy_context import DeployContext
+from samcli.commands.package.package_context import PackageContext
+from samcli.lib.providers.sam_stack_provider import is_local_path
+from samcli.lib.utils.boto_utils import get_boto_client_provider_from_session_with_config
+from samcli.lib.utils.resources import (
+    AWS_APIGATEWAY_RESTAPI,
+    AWS_APIGATEWAY_V2_API,
+    AWS_CLOUDFORMATION_STACK,
+    AWS_LAMBDA_FUNCTION,
+    AWS_LAMBDA_LAYERVERSION,
+    AWS_SERVERLESS_API,
+    AWS_SERVERLESS_APPLICATION,
+    AWS_SERVERLESS_FUNCTION,
+    AWS_SERVERLESS_HTTPAPI,
+    AWS_SERVERLESS_LAYERVERSION,
+    AWS_SERVERLESS_STATEMACHINE,
+    AWS_STEPFUNCTIONS_STATEMACHINE,
+    SYNCABLE_RESOURCES,
+)
+from samcli.yamlhelper import yaml_parse
+
+LOG = logging.getLogger(__name__)
+
+GENERAL_REMOVAL_MAP = {
+    AWS_SERVERLESS_FUNCTION: ["CodeUri", "ImageUri"],
+    AWS_SERVERLESS_LAYERVERSION: ["ContentUri"],
+    AWS_LAMBDA_LAYERVERSION: ["Content"],
+    AWS_SERVERLESS_API: ["DefinitionBody"],
+    AWS_APIGATEWAY_RESTAPI: ["BodyS3Location"],
+    AWS_SERVERLESS_HTTPAPI: ["DefinitionUri"],
+    AWS_APIGATEWAY_V2_API: ["BodyS3Location"],
+    AWS_SERVERLESS_STATEMACHINE: ["DefinitionUri"],
+    AWS_STEPFUNCTIONS_STATEMACHINE: ["DefinitionS3Location"],
+}
+
+LAMBDA_FUNCTION_REMOVAL_MAP = {
+    AWS_LAMBDA_FUNCTION: {"Code": ["ImageUri", "S3Bucket", "S3Key", "S3ObjectVersion"]},
+}
+
+
+class InfraSyncExecutor:
+    """
+    Executor for infra sync that contains skip logic when template is not changed
+    """
+
+    _build_context: BuildContext
+    _package_context: PackageContext
+    _deploy_context: DeployContext
+
+    def __init__(self, build_context: BuildContext, package_context: PackageContext, deploy_context: DeployContext):
+        """Constructs the sync for infra executor.
+
+        Parameters
+        ----------
+        build_context : BuildContext
+            BuildContext
+        package_context : PackageContext
+            PackageContext
+        deploy_context : DeployContext
+            DeployContext
+        """
+        self._build_context = build_context
+        self._package_context = package_context
+        self._deploy_context = deploy_context
+
+        self._session = Session(profile_name=self._deploy_context.profile, region_name=self._deploy_context.region)
```

Comment:
If we are instantiating the clients and not using `self._session` anymore, we don't need to keep it as a class field here.